pkg/client/kubernetes/clientmap/internal/deleg_clientmap.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap"
+)
+
+var _ clientmap.ClientMap = &delegatingClientMap{}
+
+// delegatingClientMap is a ClientMap, which will delegate calls to different ClientMaps based on
+// the type of the key (e.g. a call with keys.ForShoot() will be delegated to the ShootClientMap).
+type delegatingClientMap struct {
+	GardenClients clientmap.ClientMap
+	SeedClients   clientmap.ClientMap
+	ShootClients  clientmap.ClientMap
+	PlantClients  clientmap.ClientMap
+}
+
+// NewDelegatingClientMap constructs a new delegatingClientMap consisting of the given different ClientMaps.
+// It will panic if `gardenClientMap` is nil.
+func NewDelegatingClientMap(gardenClientMap, seedClientMap, shootClientMap, plantClientMap clientmap.ClientMap) clientmap.ClientMap {
+	if gardenClientMap == nil {
+		panic("delegatingClientMap must contain a non-nil gardenClientMap")
+	}
+
+	return &delegatingClientMap{
+		GardenClients: gardenClientMap,
+		SeedClients:   seedClientMap,
+		ShootClients:  shootClientMap,
+		PlantClients:  plantClientMap,
+	}
+}
+
+// ErrUnknownKeyType is an error that will be returned by a delegatingClientMap if the type of the given key is unknown.
+type ErrUnknownKeyType struct {
```

Comment:
```suggestion
type errUnknownKeyType struct {
```
Since it's in an `internal` package and not part of the API surface.